docs/standard/serialization/binaryformatter-migration-guide/reading-nrbf-payloads.md

```diff
@@ -0,0 +1,206 @@
+
+### Identifying BinaryFormatter payload
+
+`NrbfDecoder` provides two `StartsWithPayloadHeader` methods that allow to **check whether given stream or buffer starts with NRBF header**.
+
+Using these methods is recommended for the period of migrating payloads persisted with `BinaryFormatter` to a [different serializer](./choosing-a-serializer.md):
+
+- Check if the payload read from storage is an [NRBF](/openspecs/windows_protocols/ms-nrbf/75b9fe09-be15-475f-85b8-ae7b7558cfe5) payload.
+- If so, read it with `NrbfDecoder`, serialize it back with a new serializer and overwrite the data in the storage.
+- If not, use the new serializer to deserialize the data.
+
+### Safely reading NRBF payloads
+
+The NRBF payload consists of serialization records that represent the serialized objects and their metadata. To read the whole payload and get the root object, the user needs to call the `Decode` method.
+
+The `Decode` method returns an `SerializationRecord` instance. `SerializationRecord` is an abstract class that represents the serialization record and provides three self-describing properties: `Id`, `RecordType` and `TypeName`. It exposes one method: `public bool TypeNameMatches(Type type)` which compares the type name read from the payload (and exposed via `TypeName` property) against the specified type. This method ignores assembly names, so the users don't need to worry about type forwarding and assembly versioning. It also does not consider member names or their types (because getting this information would require type loading).
+
+```cs
+using System.Formats.Nrbf;
+
+static T Pseudocode<T>(Stream payload)
+{
+    SerializationRecord record = NrbfDecoder.Read(payload);
+    if (!record.TypeNameMatches(typeof(T))
+    {
+        throw new Exception($"Expected the record to match type name `{typeof(T).AssemblyQualifiedName}`, but got `{record.TypeName.AssemblyQualifiedName}`."
+    }
+}
+```
+
+There is more than a dozen of different serialization [record types](/openspecs/windows_protocols/ms-nrbf/954a0657-b901-4813-9398-4ec732fe8b32), but this library provides a set of abstractions, so the users need to learn only a few of them:
+
+- `PrimitiveTypeRecord<T>` that describes all primitive types natively supported by the NRBF (`string`, `bool`, `byte`, `sbyte`, `char`, `short`, `ushort`, `int`, `uint`, `long`, `ulong`, `float`, `double`, `decimal`, `TimeSpan` and `DateTime`) and exposes the value `Value` getter. `PrimitiveTypeRecord<T>` derives from non-generic `PrimitiveTypeRecord` which also exposes the raw value via `Value` property, but it's returned as `object` (which introduces boxing for value types).
```

Comment:
I'd use a colon instead of "that" for all of these.

```suggestion
- `PrimitiveTypeRecord<T>`: describes all primitive types natively supported by the NRBF (`string`, `bool`, `byte`, `sbyte`, `char`, `short`, `ushort`, `int`, `uint`, `long`, `ulong`, `float`, `double`, `decimal`, `TimeSpan` and `DateTime`) and exposes the value `Value` getter. `PrimitiveTypeRecord<T>` derives from non-generic `PrimitiveTypeRecord` which also exposes the raw value via `Value` property, but it's returned as `object` (which introduces boxing for value types).
```